File: python/VCFImputedAndUnknown.py
from __future__ import annotations

# coding: utf-8
# VCFImputedAndUnknown.py
# 片親がimputeされていて、もう片親はUnknown
# まず片親

from VCF import VCFSmall
from VCFGeno import VCFGenoBase
from GenoRecord import GenoRecord
from VCFFamily import *
from VCFImputable import VCFImputable
from ProgenyImputerByOneParent import *
from ProgenyImputer import *
from Map import *


#################### VCFImputedAndUnknown ####################

class VCFImputedAndUnknown(VCFImputable):
	def __init__(self, samples: list[str], records: list[VCFFamilyRecord],
					ref_haps: list[list[int]], is_mat_imputed: bool,
					map_: Map, vcf: VCFSmall):
		VCFImputable.__init__(self, samples, vcf)
		self.records: list[VCFFamilyRecord] = records
		self.prog_imputer = ProgenyImputerByOneParent(records, ref_haps,
														is_mat_imputed,
														map_, 0.01)
	
	##### virtual methods for VCFGenoBase #####
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_records(self) -> list[GenoRecord]:
		return [ r for r in self.records ]
	
	##### virtual methods for VCFFamilyBase #####
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	##### virtual methods for VCFImputable #####
	def impute(self) -> None:
		# 最初の後代をimpute
		self.prog_imputer.impute(0)

File: src/VCFImputedAndUnknown.cpp
#include <cmath>
#include "../include/VCFImputedAndUnknown.h"
#include "../include/common.h"

using namespace std;

VCFImputedAndUnknown::VCFImputedAndUnknown(const STRVEC& s,
							const std::vector<VCFFamilyRecord *>& rs,
							const std::vector<std::vector<int>>& ref_hs,
							bool is_mat_imputed,
							const Map& map_, double w, const VCFSmall *vcf) :
					VCFImputable(s, vcf), records(rs), ref_haps(ref_hs),
					prog_imputer(records, ref_hs, is_mat_imputed, map_, w),
					parent_imputer(records, ref_hs, is_mat_imputed, map_, w),
					progs_imputer(records, map_, w) { }

VCFImputedAndUnknown::~VCFImputedAndUnknown() {
	Common::delete_all(records);
}

size_t VCFImputedAndUnknown::amount() const {
	const size_t	M = ref_haps[0].size();
	const size_t	NH = ref_haps.size();
	const size_t	R = NH*NH*20;
	return R * M;
}

void VCFImputedAndUnknown::impute() {
	prog_imputer.impute(0);
	parent_imputer.impute();
	for(size_t j = 1; j < num_progenies(); ++j) {
		progs_imputer.impute(j);
	}
}
